### PR TITLE
Improve resiliency of GitHubStatusPush

### PR DESCRIPTION
**Pull request by steven-johnson**
If you attempt to use GitHubStatusPush with a builder that specifies multiple codebases, you can currently get runtime failures:
- the existing code assumes that the [0]'th entry of sourcestamps always has valid 'branch' property (it might not, if the codebase is one that is pulled without a branch specification)
- the existing code assumes that *all* entries of sourcestamps have a valid 'revision' property (same reason as above)

This revises the code to be more defensive, and skip the update attempts for sourcestamps that have no revision data (logging the skip in verbose mode).

### CODE REVIEW

**Comment by tardyp** [comment]
I would prefer to get some unit tests for those use cases

**Comment by steven-johnson** [comment]
Thanks. It will probably take me a few days to add the test cases.

File: master/buildbot/reporters/github.py [hunk 1]
```diff
@@ -126,50 +126,53 @@ def send(self, build):
...
-        project = sourcestamps[0]['project']
+        for sourcestamp in sourcestamps:
+            project = sourcestamp['project']
 
-        branch = props['branch']
-        m = re.search(r"refs/pull/([0-9]*)/merge", branch)
-        if m:
-            issue = m.group(1)
-        else:
+            # If branch is None, so is issue
```
Comment (sa2ajj) [inline master/buildbot/reporters/github.py]: this line probably does not belong to the change

File: master/buildbot/reporters/github.py [hunk 2]
```diff
@@ -126,50 +126,53 @@ def send(self, build):
...
             issue = None
+            if 'branch' in props:
+                m = re.search(r"refs/pull/([0-9]*)/merge", props['branch'])
+                if m:
+                    issue = m.group(1)
+
+            repoOwner = None
```
Comment (sa2ajj) [inline master/buildbot/reporters/github.py]: i understand we follow Twisted naming scheme in a lot of places, there's no need to do it here: `repo_owner` would be just fine

File: master/buildbot/reporters/github.py [hunk 3]
```diff
@@ -126,50 +126,53 @@ def send(self, build):
...
+            repoName = None
```
Comment (sa2ajj) [inline master/buildbot/reporters/github.py]: ditto (wrt `repo_name`)

File: master/buildbot/reporters/github.py [hunk 4]
```diff
@@ -126,50 +126,53 @@ def send(self, build):
...
-
-        for sourcestamp in sourcestamps:
             sha = sourcestamp['revision']
+            # If the scheduler specifies multiple codebases, don't bother updating
+            # the ones for which there is no revision
+            if not sha:
+                log.msg(
+                    'Skipped status update for codebase {codebase}, '
+                    'context "{context}", issue {issue}.'.format(
+                        codebase=sourcestamp['codebase'], issue=issue, context=context))
```
Comment (sa2ajj) [inline master/buildbot/reporters/github.py]: this line **could** be formatted as one parameter per line w/ `context=...` having a trailing `,` for easier future reviews

File: master/buildbot/reporters/github.py (hunk 3 above)
Comment (steven-johnson) [reply master/buildbot/reporters/github.py]: These are pre-existing names (my change just shuffled them around).

**Comment by krf** [comment]
@steven-johnson I'd be interested in getting this fixed, esp. wrt #4105 -- any incentives to pick this up again?

**Comment by steven-johnson** [comment]
> @steven-johnson I'd be interested in getting this fixed, esp. wrt #4105 -- any incentives to pick this up again?

I don't have the bandwidth to add the unit tests at this time, unfortunately. I'd welcome someone taking over this patch and finishing it, though.

**Comment by steven-johnson** [comment]
FYI: after a long-overdue upgrade of our buildbot from 1.x to 2.8, I find that this failure has moved from "once in a while" to "pretty much every single time" -- GitHubStatusPush no longer seems to work *at all* for our setup; the log files indicate that ~every usage of it fails with some variation of

```
        Traceback (most recent call last):
          File "/path/to/python/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
            result = g.send(result)
          File "/path/to/python/lib/python3.7/site-packages/buildbot/reporters/http.py", line 80, in getMoreInfoAndSen
d
            yield self.send(build)
          File "/path/to/python/lib/python3.7/site-packages/twisted/internet/defer.py", line 1613, in unwindGenerator
            return _cancellableInlineCallbacks(gen)
          File "/path/to/python/lib/python3.7/site-packages/twisted/internet/defer.py", line 1529, in _cancellableInli
neCallbacks
            _inlineCallbacks(None, g, status)
        --- <exception caught here> ---
          File "/path/to/python/lib/python3.7/site-packages/buildbot/reporters/http.py", line 80, in getMoreInfoAndSen
d
            yield self.send(build)
          File "/path/to/python/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
            result = g.send(result)
          File "/path/to/python/lib/python3.7/site-packages/buildbot/reporters/github.py", line 132, in send
            branch = props['branch']
          File "/path/to/python/lib/python3.7/site-packages/buildbot/process/properties.py", line 140, in __getitem__
            rv = self.properties[name][0]
        builtins.KeyError: 'branch'
```

File: master/buildbot/reporters/github.py (hunk 1 above)
Comment (p12tic) [reply master/buildbot/reporters/github.py:135]: Fixed now.

File: master/buildbot/reporters/github.py (hunk 2 above)
Comment (p12tic) [reply master/buildbot/reporters/github.py:142]: Fixed now.

File: master/buildbot/reporters/github.py (hunk 3 above)
Comment (p12tic) [reply master/buildbot/reporters/github.py:143]: Fixed now.

**Comment by p12tic** [comment]
This as been merged as part of #5316, removing stale label.
